crates/hir/src/lib.rs

```diff
@@ -5972,6 +5972,63 @@ impl Layout {
         }
     }
 
+    pub fn tail_padding(&self, field_size: &mut impl FnMut(usize) -> Option<u64>) -> Option<u64> {
+        match self.0.fields {
+            layout::FieldsShape::Primitive => None,
+            layout::FieldsShape::Union(_) => None,
+            layout::FieldsShape::Array { stride, count } => count.checked_sub(1).and_then(|tail| {
+                let tail_field_size = field_size(tail as usize)?;
+                let offset = stride.bytes() * tail;
+                self.0.size.bytes().checked_sub(offset)?.checked_sub(tail_field_size)
+            }),
+            layout::FieldsShape::Arbitrary { ref offsets, ref memory_index } => {
+                let tail = memory_index.last_index()?;
+                let tail_field_size = field_size(tail.0.into_raw().into_u32() as usize)?;
+                let offset = offsets.get(tail)?.bytes();
+                self.0.size.bytes().checked_sub(offset)?.checked_sub(tail_field_size)
+            }
+        }
+    }
+
+    pub fn largest_padding(
+        &self,
+        field_size: &mut impl FnMut(usize) -> Option<u64>,
+    ) -> Option<u64> {
+        match self.0.fields {
+            layout::FieldsShape::Primitive => None,
+            layout::FieldsShape::Union(_) => None,
+            layout::FieldsShape::Array { stride, count } => (0..count)
+                .map(|i| (i, stride.bytes() * i))
+                .chain(std::iter::once((count, self.0.size.bytes())))
+                .tuple_windows()
+                .filter_map(|((i, start), (_, end))| {
+                    let size = field_size(i as usize)?;
+                    end.checked_sub(start)?.checked_sub(size)
+                })
+                .max(),
```

Comment:
No need to loop here, the elements should all have the same padding right?
```suggestion
            layout::FieldsShape::Array { stride: _, count: 0..=1 }  => None,
            layout::FieldsShape::Array { stride, count: _ }  => {
                    let size = field_size(0 as usize)?;
                    stride.bytes().checked_sub(size)
            },
```

Reply:
Yes, I mindlessly wrote those lines with just muscle memory 😅